modules/delegates.js

```diff
@@ -353,112 +353,160 @@ __private.decryptPassphrase = function(encryptedPassphrase, password) {
-								math === '+' &&
-								(delegates != null && delegates.indexOf(publicKey) !== -1)
-							) {
-								return setImmediate(
-									cb,
-									`Failed to add vote, delegate "${
-										account.username
-									}" already voted for`
-								);
-							}
-
-							if (
-								math === '-' &&
-								(delegates === null || delegates.indexOf(publicKey) === -1)
-							) {
-								return setImmediate(
-									cb,
-									`Failed to remove vote, delegate "${
-										account.username
-									}" was not voted for`
-								);
-							}
-
-							return setImmediate(cb);
-						},
-						tx
-					);
-				},
-				err => {
-					if (err) {
-						return setImmediate(cb, err);
-					}
-
-					const totalVotes = existingVotes + additions - removals;
+			voteAddress = modules.accounts.generateAddressByPublicKey(votePublicKey);
+		} catch (e) {
+			library.logger.error(e.stack);
+			return setImmediate(cb, e);
+		}
+		return {
+			action,
+			address: voteAddress,
+		};
+	});
 
-					if (totalVotes > constants.activeDelegates) {
-						const exceeded = totalVotes - constants.activeDelegates;
+	async.waterfall(
+		[
+			// get all the addresses (converted from public keys) the sender has voted for. Confirmed or unconfirmed based on state parameter.
+			function getExistingVotedAddresses(waterfallCb) {
+				modules.accounts.getAccount(
+					{ publicKey: senderPublicKey },
+					(err, account) => {
+						if (err) {
+							return setImmediate(waterfallCb, err);
+						}
+
+						if (!account) {
+							return setImmediate(waterfallCb, 'Account not found');
+						}
+
+						const delegates =
+							state === 'confirmed' ? account.delegates : account.u_delegates;
+						const votedAddresses = Array.isArray(delegates)
+							? delegates.map(delegate =>
+									modules.accounts.generateAddressByPublicKey(delegate)
+								)
+							: [];
+
+						return setImmediate(waterfallCb, null, votedAddresses);
+					},
+					tx
+				);
+			},
+			// Validate votes in the transaction by checking that sender is not voting for an account already, and also that sender is not unvoting an account it did not vote before.
+			function validateVotes(existingVotedAddresses, waterfallCb) {
+				modules.accounts.getAccounts(
+					{
+						address: actionWithAddresses.map(({ address }) => address),
+						isDelegate: 1,
+					},
+					(err, votesAccounts) => {
+						if (err) {
+							return setImmediate(waterfallCb, err);
+						}
+
+						if (
+							!votesAccounts ||
+							votesAccounts.length < actionWithAddresses.length
+						) {
+							library.logger.error(
+								'Delegates with addresses not found',
+								_.differenceWith(
+									actionWithAddresses,
+									votesAccounts,
+									(
+										{ address: addressFromTransaction },
+										{ addressFromAccount }
+									) => addressFromTransaction === addressFromAccount
+								)
+							);
+							return setImmediate(waterfallCb, 'Delegate not found');
+						}
```

Comment:
There are two different steps of validation: 

1. The votes (or public keys) appears valid delegates 
2. Then validate individual vote

So the flow should be implemented as: 

1. Parse all votes to strip the actions 
2. Convert all public keys to addresses 
3. Check if any of those are not valid delegate 
4. Fetch all existing votes from that account 
5. Validate individual vote 

I believe that's the more efficient and effective flow to validate votes. 

Currently the order is implemented as 1, 2, 4, 3, 5.

Reply:
Why do you say its more efficient? I agree that the flow seems more consistent in the order you described it. But 4 fetches 1 record, and 3 will fetch 33 records. So, it's better to perform step 4  because it requires less I/O.

Reply:
As if something failed in step 3, we don't need to go for step 4.

Reply:
It's a matter of preference. And I prefer the current approach. Let me know if you strongly feel about this. If so, we can discuss it further.

Reply:
I still feel that its not matter of the preference. Means its not a syntax or js feature usage. Its a business flow, if you feel that above steps are the right order, then we must go with those. 

If you feel above steps are not in correct order, then we can discuss further.

Reply:
The well-informed decision about the order would have to rely on studies - is it more probable to receive a transaction that exceeds the votes limit (4. Fetch all existing votes from that account) or votes for accounts not being delegates (3. Check if any of those are not valid delegate). 
My intuition basing on the current situation says the exceeding the votes limit is more common situation. It has to be done first then to avoid unnecessary following steps. 
Then the order implemented by Usman is valid.